drivers/can/can_stm32h7.c

```diff
@@ -0,0 +1,304 @@
+									       \
+PINCTRL_DT_INST_DEFINE(inst);						       \
+									       \
+static const struct can_stm32h7_config can_stm32h7_cfg_##inst = {              \
+	.msg_sram = (struct can_mcan_msg_sram *)                               \
+			DT_INST_REG_ADDR_BY_NAME(inst, message_ram),           \
+	.config_irq = config_can_##inst##_irq,                                 \
+	.mcan_cfg = {                                                          \
+		.can = (struct can_mcan_reg *)                                 \
+			DT_INST_REG_ADDR_BY_NAME(inst, m_can),                 \
+		.bus_speed = DT_INST_PROP(inst, bus_speed),                    \
+		.sjw = DT_INST_PROP(inst, sjw),                                \
+		.sample_point = DT_INST_PROP_OR(inst, sample_point, 0),        \
+		.prop_ts1 = DT_INST_PROP_OR(inst, prop_seg, 0) +               \
+			    DT_INST_PROP_OR(inst, phase_seg1, 0),              \
+		.ts2 = DT_INST_PROP_OR(inst, phase_seg2, 0),                   \
+		.bus_speed_data = DT_INST_PROP(inst, bus_speed_data),          \
+		.sjw_data = DT_INST_PROP(inst, sjw_data),                      \
+		.sample_point_data =                                           \
+			DT_INST_PROP_OR(inst, sample_point_data, 0),           \
+		.prop_ts1_data = DT_INST_PROP_OR(inst, prop_seg_data, 0) +     \
+				DT_INST_PROP_OR(inst, phase_seg1_data, 0),     \
+		.ts2_data = DT_INST_PROP_OR(inst, phase_seg2_data, 0),         \
+		.tx_delay_comp_offset =                                        \
+			DT_INST_PROP(inst, tx_delay_comp_offset)               \
+	},                                                                     \
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),                          \
+};
+
+#else /* CONFIG_CAN_FD_MODE */
+
+#define CAN_STM32H7_CFG_INST(inst)                                             \
+									       \
+PINCTRL_DT_INST_DEFINE(inst);						       \
+									       \
+static const struct can_stm32h7_config can_stm32h7_cfg_##inst = {              \
+	.msg_sram = (struct can_mcan_msg_sram *)                               \
+			DT_INST_REG_ADDR_BY_NAME(inst, message_ram),           \
+	.config_irq = config_can_##inst##_irq,                                 \
+	.mcan_cfg = {                                                          \
+		.can = (struct can_mcan_reg *)                                 \
+			DT_INST_REG_ADDR_BY_NAME(inst, m_can),                 \
+		.bus_speed = DT_INST_PROP(inst, bus_speed),                    \
+		.sjw = DT_INST_PROP(inst, sjw),                                \
+		.sample_point = DT_INST_PROP_OR(inst, sample_point, 0),        \
+		.prop_ts1 = DT_INST_PROP_OR(inst, prop_seg, 0) +               \
+			    DT_INST_PROP_OR(inst, phase_seg1, 0),              \
+		.ts2 = DT_INST_PROP_OR(inst, phase_seg2, 0),                   \
+	},                                                                     \
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),                          \
+};
```

Comment:
there's a significant amount of common code here, a single device structure definitions macro can be used. Non-common parts should be made conditional.

Reply:
@gmarull I'm not sure how to do that, can't use an #ifdef inside of a #define—can you point me at an example of how to clean this up?

Reply:
You can use a macro which definition would be conditional to CONFIG_CAN_FD_MODE statu

Reply:
Isn't that what's in there now?

Reply:
I've cleaned this up, much better.